File: vibecheck-core/tests/fixtures/lru_cache/gemini.js
// LRU Cache
// - doubly linked list + map for O(1) ops
// - async fetch helper for cache-miss loading
// - fixed capacity with automatic eviction

class Node {
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

class Cache {
    constructor(cap) {
        this.cap = cap;
        this.items = new Map();
        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    get(key) {
        const found = this.items.has(key);
        const node = found ? this.items.get(key) : null;
        if (!found) return -1;
        this._detach(node);
        this._attach(node);
        return node.value;
    }

    put(key, value) {
        // - update if key exists already
        // - evict oldest when at capacity
        // - insert new entry at front
        const exist = this.items.has(key);
        if (exist) {
            const node = this.items.get(key);
            node.value = value;
            this._detach(node);
            this._attach(node);
            return;
        }

        const full = this.items.size >= this.cap;
        if (full) {
            const { key: oldKey } = this.tail.prev;
            this._detach(this.tail.prev);
            this.items.delete(oldKey);
        }

        const node = new Node(key, value);
        this.items.set(key, node);
        this._attach(node);
    }

    async fetch(key, loader) {
        // - return cached value if present
        // - await loader on cache miss
        // - store loaded result in cache
        const hit = this.items.has(key);
        if (hit) return this.get(key);

        const value = await loader(key);
        this.put(key, value);
        return value;
    }

    _detach(node) {
        const { prev, next } = node;
        prev.next = next;
        next.prev = prev;
    }

    _attach(node) {
        const { next: first } = this.head;
        this.head.next = node;
        node.prev = this.head;
        node.next = first;
        first.prev = node;
    }

    size() {
        return this.items.size;
    }

    keys() {
        const result = [];
        let curr = this.head.next;
        while (curr !== this.tail) {
            const { key } = curr;
            result.push(key);
            curr = curr.next;
        }
        return result.length > 0 ? result : [];
    }

    clear() {
        this.items.clear();
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }
}

module.exports = { Cache, Node };

File: vibecheck-core/tests/fixtures/lru_cache/gemini.go
package cache

import "sync"

// LRU Cache
// - O(1) get and put with map + linked list
// - mutex for concurrent safety
// - defer unlock pattern throughout

type Node struct {
	key   int
	value int
	prev  *Node
	next  *Node
}

type Cache struct {
	items map[int]*Node
	head  *Node
	tail  *Node
	cap   int
	mu    sync.Mutex
}

func New(cap int) *Cache {
	head := &Node{}
	tail := &Node{}
	head.next = tail
	tail.prev = head
	return &Cache{
		items: make(map[int]*Node),
		head:  head,
		tail:  tail,
		cap:   cap,
	}
}

func (c *Cache) Get(key int) (int, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	node, found := c.items[key]
	// - return zero value on miss
	// - move to front on hit
	// - return stored value
	if !found {
		return 0, false
	}
	c.detach(node)
	c.attach(node)
	return node.value, true
}

func (c *Cache) Put(key int, value int) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// - update existing entry if present
	// - evict oldest when at capacity
	// - attach new entry at head
	if node, ok := c.items[key]; ok {
		node.value = value
		c.detach(node)
		c.attach(node)
		return
	}

	full := len(c.items) >= c.cap
	if full {
		old := c.tail.prev
		c.detach(old)
		delete(c.items, old.key)
	}

	node := &Node{key: key, value: value}
	c.items[key] = node
	c.attach(node)
}

func (c *Cache) detach(node *Node) {
	prev := node.prev
	next := node.next
	prev.next = next
	next.prev = prev
}

func (c *Cache) attach(node *Node) {
	first := c.head.next
	c.head.next = node
	node.prev = c.head
	node.next = first
	first.prev = node
}

func (c *Cache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.items)
}

func (c *Cache) Keys() []int {
	c.mu.Lock()
	defer c.mu.Unlock()

	result := make([]int, 0, len(c.items))
	curr := c.head.next
	for curr != c.tail {
		result = append(result, curr.key)
		curr = curr.next
	}
	empty := len(result) == 0
	if empty {
		return nil
	}
	return result
}
